### PR TITLE
Check If User Inputted Coordinates are Valid

### PR DESCRIPTION
**Pull request by hellosweta**
This commit prompts user until they enter a valid input. Game quits upon valid input.

### CODE REVIEW

File: lib/game.rb [hunk 1]
```diff
@@ -1,17 +1,63 @@
...
+
   def self.run(*args)
-    Game.new.run(*args)
+    Game.new(*args).run
   end
 
-  def run(output)
-    output.puts empty_board
+  def run
+    coordinates = start_game
```
Comment (thorncp) [inline lib/game.rb]: One of the design guides is
> Don't use local variables, make methods instead that capture that logic (and memoize if necessary)

Perhaps a method further down can update a `@coordinates` variable?

File: spec/coordinates_spec.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+require "spec_helper"
+require_relative "../lib/coordinates"
+
+RSpec.describe Coordinates do
+  describe "valid?" do
+    it "returns false when row is not valid" do
```
Comment (thorncp) [inline spec/coordinates_spec.rb]: This description is kinda like using the word in the definition of the word. What are the characteristics that make the row invalid?

E.g. since the grid is visual, a cell that's beyond the top/right/bottom/left is invalid. What do you think about making specs specifically to test those edge cases?

File: spec/coordinates_spec.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+    it "returns false when column is not valid" do
+      input = "a4"
+
+      coordinates = Coordinates.new(input)
+
+      expect(coordinates.valid?).to be false
+    end
+
+    it "returns true when row and column are valid" do
```
Comment (thorncp) [inline spec/coordinates_spec.rb]: And here, what characteristics make a coordinate valid? I don't think you need to exhaust every possible valid cell, but I think it's worth testing upper and lower case rows.

File: spec/game_spec.rb [hunk 4]
```diff
@@ -13,12 +13,29 @@
...
+
+    it "exits when user inputs a valid coordinate" do
+      mock_stdout = StringIO.new
+      mock_stdin = StringIO.new
+      allow(mock_stdin).to receive(:gets).and_return("A1")
 
-      expect(output.string).to eq(empty_board)
+      expect {
+        Game.run(mock_stdout, mock_stdin)
+      }.to exit
```
Comment (thorncp) [inline spec/game_spec.rb]: Is `exit` an RSpec thing? My expectation, without knowing, is that this is Ruby's `exit` that will halt the spec suite.

File: lib/game.rb [hunk 5]
```diff
@@ -1,17 +1,63 @@
...
+  end
+
+  def retry_until_valid(input)
+    until valid_input?(input)
+      input = reprompt_input
+    end
+  end
+
+  def valid_input?(input)
+    false if input.length != 2
```
Comment (thorncp) [inline lib/game.rb]: This seems to be a partial validation of the coordinates. Could this be moved into the `Coordinates` class's `valid?` method?

File: lib/game.rb [hunk 6]
```diff
@@ -1,17 +1,63 @@
...
+  def run
+    coordinates = start_game
+    retry_until_valid(coordinates)
+    SystemExit
   end
 
   private
 
+  def start_game
+    @std_out.puts game_start_message
```
Comment (thorncp) [inline lib/game.rb]: What do you think about using `attr_reader` for these?

File: lib/game.rb [hunk 7]
```diff
@@ -1,17 +1,63 @@
...
+    "Where would you like to play?"
+  end
+
+  def instructions_prompt
+    'Enter a row letter and column number. For Example: "A1".'
+  end
+
+  def retry_until_valid(input)
+    until valid_input?(input)
+      input = reprompt_input
```
Comment (thorncp) [inline lib/game.rb]: The word "prompt" is about outputting, which makes the return value of an input a little confusing.

File: lib/game.rb (hunk 1 above)
Comment (hellosweta) [reply lib/game.rb]: I like that idea, what do you think about 04146cb4ca666e71809175cd780a7e45b0f20859

File: lib/game.rb [hunk 8]
```diff
@@ -1,17 +1,71 @@
...
   def self.run(*args)
-    Game.new.run(*args)
+    Game.new(*args).run
   end
 
-  def run(output)
-    output.puts empty_board
+  def run
+    start_game
+    retry_until_valid
```
Comment (thorncp) [inline lib/game.rb]: In the context of the `run` method, I'm wondering _what_ is being retried here. What do you think about a higher level name for this method to fit in better with `start_game` and `end_game`? Maybe something along the lines of `get_user_coordinates`?

File: lib/coordinates.rb [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  def initialize(user_input)
+    @user_input = user_input
+  end
+
+  def valid?
+    return false unless correct_length?
+    row_valid? && column_valid?
+  end
+
+  def correct_length?
```
Comment (thorncp) [inline lib/coordinates.rb:14]: Could this be private?

File: lib/coordinates.rb [hunk 10]
```diff
@@ -0,0 +1,44 @@
+class Coordinates
+  attr_reader :user_input, :row, :column
+
+  def initialize(user_input)
+    @user_input = user_input
+  end
+
+  def valid?
+    return false unless correct_length?
```
Comment (thorncp) [inline lib/coordinates.rb]: Could this be combined with the conditional on the next line? e.g `correct_length? && row_valid? && column_valid?`

File: lib/game.rb [hunk 11]
```diff
@@ -1,17 +1,71 @@
...
+      reprompt_input
+    end
+  end
+
+  def valid_input?
+    coordinates.valid?
+  end
+
+  def set_coordinates_from_user_input
+    @coordinates = Coordinates.new(std_in.gets)
```
Comment (thorncp) [inline lib/game.rb]: When you run the game for real, can you see it get valid input and exit? `gets` includes the ending newline, so a value coming in here looks like `"A2\n"`, which I think won't pass your validation checks. Generally, scripts in Ruby will do `gets.chomp` to swallow the newline.

File: spec/coordinates_spec.rb (hunk 2 above)
Comment (hellosweta) [reply spec/coordinates_spec.rb]: Good call out. Done in 4035b4a8b4432bfb9dd86248e85b06289a131d18. Do you have a recommendation to test a "row before the first" is invalid. What's before "A"? hehe

File: spec/coordinates_spec.rb (hunk 3 above)
Comment (hellosweta) [reply spec/coordinates_spec.rb]: What do you think about this change 4035b4a8b4432bfb9dd86248e85b06289a131d18

File: spec/game_spec.rb (hunk 4 above)
Comment (hellosweta) [reply spec/game_spec.rb]: Thanks for pointing this out! Updated to catch the error. c522627b9df6321dd5fe822111df2b01218d4ee5

File: lib/game.rb (hunk 7 above)
Comment (hellosweta) [reply lib/game.rb]: Good call this is now updated

File: spec/coordinates_spec.rb (hunk 2 above)
Comment (thorncp) [reply spec/coordinates_spec.rb]: Yeah, before first row is weird. I used `@2`, because `@` is before `A` in ASCII, but then the test no longer made sense as "before the first row" and I changed it to being about malformed data. That's similar to the spec you have with `a43`

File: lib/game.rb (hunk 11 above)
Comment (hellosweta) [reply lib/game.rb]: I swear this was working before! but you're right I needed to add chomp!

File: lib/coordinates.rb (hunk 10 above)
Comment (hellosweta) [reply lib/coordinates.rb]: hmm, it can't right now because `column_valid` requires a string that's at least two characters long.

File: lib/game.rb (hunk 8 above)
Comment (hellosweta) [reply lib/game.rb]: good point. I like how you're reading the method steps, and how the method name stands out when you do that. Updating.

File: spec/coordinates_spec.rb (hunk 2 above)
Comment (hellosweta) [reply spec/coordinates_spec.rb]: hehe I like that you used `@`. I'm going to omit "before first row" then. What do you think about the edge cases we're testing now?

File: lib/coordinates.rb (hunk 10 above)
Comment (thorncp) [reply lib/coordinates.rb]: If it's not the correct length, I think it would short circuit before calling `column_valid?`.

File: lib/game.rb (hunk 8 above)
Comment (hellosweta) [reply lib/game.rb]: I also like that the user doesn't have to think about what's going on under the hood -- the "retry" part is not important!

File: spec/coordinates_spec.rb [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  describe "valid?" do
+    it "returns false when user entered row is below last row" do
+      input = "d1"
+
+      coordinates = Coordinates.new(input)
+
+      expect(coordinates.valid?).to be false
+    end
+
+    it "returns false when user entered column is above first column" do
```
Comment (thorncp) [inline spec/coordinates_spec.rb]: above -> before ?

File: spec/coordinates_spec.rb (hunk 2 above)
Comment (thorncp) [reply spec/coordinates_spec.rb]: They look good!

File: lib/coordinates.rb (hunk 10 above)
Comment (hellosweta) [reply lib/coordinates.rb]: teehee you're right!

**Review by thorncp — APPROVED** [review]
Looks good. The `Game` class has taken on quite a few responsibilities now - it might be time to break it apart.

**Comment by hellosweta** [comment]
> Looks good. The Game class has taken on quite a few responsibilities now - it might be time to break it apart.

You make a good point! Next refactor!
Also, is this more in line with how to put PRs up with red, green, refactor commits in one PR? I think it worked well, and I wanted to make sure it was working for you.

**Comment by thorncp** [comment]
> Also, is this more in line with how to put PRs up with red, green, refactor commits in one PR? I think it worked well, and I wanted to make sure it was working for you.

Yep, it worked for me 👍
